### ISSUE TITLE
Zeus Test intermittently exits with code 2

### ISSUE THREAD

**Issue by RobinClowers** [burke/zeus#469]
## Description of Problem

Sometimes when running zeus test (or rspec) I get an immediate exit with code 2. It seems like a timing issue, if I run a spec right after I save a file it seems to happen.
## System details
- **`uname -a`**: Darwin robin.local 13.2.0 Darwin Kernel Version 13.2.0: Thu Apr 17 23:03:13 PDT 2014; root:xnu-2422.100.13~1/RELEASE_X86_64 x86_64
- **`ruby -v`**: ruby 2.1.1p76 (2014-02-24 revision 45161) [x86_64-darwin13.0]
## Steps to Reproduce

1) `zeus start` in a new rails project

2) Run `zeus test` from a script, something like this:

```
for i in {1..10}
do
  zeus rspec $1
done
```
## Observed Behavior
- One of the test runs will exit with code 2
## Expected Behavior
- Should run tests

**Comment by tfrey7** [comment]
My team also frequently encounters this bug.

**Comment by Adam262** [comment]
I encounter this bug very frequently when in selenium / capybara testing.  I usually need to run a `<zeus rspec>` test two or three times, even if I wait a few seconds after manually saving.

**Comment by jooshbzm** [comment]
I too am running into this issue

**Comment by latortuga** [comment]
@burke any thoughts on this? I'm not super familiar with go or the internal workings of the go portion of zeus. The issue description strikes me as a race condition.

**Comment by burke** [comment]
Yeah, it seems like zeusclient is crashing. When it connects to the master, it sets up a bunch of terminal settings, and resets them before it exits. If it exits non-gracefully, it doesn't get a chance to reset the terminal settings, resulting in the screwed up line breaks and stuff from the screenshot above.

**You can fix your terminal by running `reset` or `stty sane`.**

Definitely sounds like a race condition crash of some sort. It almost certainly involves a bug in zeusclient somehow (ie. a bug completely on the master/slave side should be incapable of messing up the client terminal).

If I were going to dig into this, I'd `strace` the `zeus test ....` (ie. `strace -tt -T -f zeus test ...` or whatever) and see how far it gets. Maybe easier, given how much syscall chatter I'd expect to see, would be to just instrument the startup code in `cmd/zeus` and `zeusclient` with a bunch of `fmt.Println` calls, to see exactly how far along the binary makes it before crashing. Bugs in `zeusclient` would tend not to be nearly as crazy or nuanced as those in the master process.

**Comment by aaronjensen** [comment]
@burke what would an equivalent be for osx? probably dtruss, but when I try `dtruss zeus test spec/lib/time_spec.rb` I don't see rspec's output, so I don't really have any way of knowing if it crashed

**Comment by tonekk** [comment]
My workaround right now is to put
`ze () { zeus $@; stty sane }`
in my `.zshrc` (or `.bashrc`), then use **ze** command instead of **zeus**.

That prevents the terminal fuckup at least

**Comment by brandur** [comment]
A slightly modified version of @tonekk's script above that also works for Zeus commands with multiple parameters like `zeus rspec spec/endpoints` (note the added quotes around `zeus` and `ze` invocations):

``` sh
ze () { ARGS=$@; zeus "$@"; ZE_EC=$?; stty sane; if [ $ZE_EC = 2 ]; then ze "$ARGS"; fi }
```

Thanks for the nice solution! I'm glad I googled this because this problem was the one thing driving me nuts on otherwise perfect piece of software.

**Comment by aaronjensen** [comment]
Arguably even better:

```bash
zeus () { ARGS=$@; command zeus "$@"; ZE_EC=$?; stty sane; if [ $ZE_EC = 2 ]; then zeus "$ARGS"; fi }
```

That way you can still say `zeus` and it'll work w/ things like vim-vroom out of the box (as long as your functions are loaded)

**Comment by smit1625** [comment]
For any friends who use [Fish Shell](http://fishshell.com/):

    function zeus; set ARGS $argv; command zeus "$argv"; set ZE_EC $status; stty sane; if $ZE_EC = 2; zeus "$ARGS"; end; end

This is based on the solution provided by @aaronjensen.

**Comment by emilsoman** [comment]
I had to modify @aaronjensen 's function a bit to get it working for me in zsh, when passing multiple arguments like `zeus rspec spec/models/user_spec.rb` : 

```
zeus () {
  ARGS=$@;
  command zeus "$@";
  ZE_EC=$?;
  if [ $ZE_EC = 2 ]; then
    stty sane;
    zeus ${=ARGS};
  fi
}
```

**Comment by TheZoq2** [comment]
Is there an update on this?
It seems to still happen and unlike things said previously, it doesn't only happen when it exits with non-zero status.

The workarounds work but as far as I can tell, it doesn't work when you want to pipe the output to something like grep or something similar.

**Comment by prpetten** [comment]
This is usually happening to me when I execute test scripts, so I've done the following.  There is probably a more efficient method of checking the return code of `super` and only running ``` `stty sane` ``` when the return code is non-zero.
```ruby
# custom_plan.rb
require 'zeus/rails'

class CustomPlan < Zeus::Rails
  def test
    super
    `stty sane`
  end
end

Zeus.plan = CustomPlan.new
```
